### PR TITLE
Refactor `_ParzenEstimator`

### PR DESCRIPTION
**Pull request by contramundum53**
Redo #4183 with correct merge.

### CODE REVIEW

**Comment by contramundum53** [comment]
@c-bata @not522 Could you review this PR again?

**Review by not522 — CHANGES_REQUESTED** [review]
Basically LGTM. Could you check my comment?

File: optuna/samplers/_tpe/probability_distributions.py [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+
+EPS = 1e-12
+
+
+def _normal_cdf(
+    x: Union[float, np.ndarray], mu: Union[float, np.ndarray], sigma: Union[float, np.ndarray]
+) -> np.ndarray:
+    return 0.5 * (1 + _erf.erf((x - mu) / np.maximum(np.sqrt(2) * sigma, EPS)))
+
+
```
Comment (not522) [inline optuna/samplers/_tpe/probability_distributions.py:44]: This function is not used.

```suggestion
```

**Review by not522 — APPROVED** [review]
LGTM except lint.

File: optuna/samplers/_tpe/_truncnorm.py [hunk 2]
```diff
@@ -207,10 +207,10 @@ def rvs(
     b: np.ndarray,
     loc: Union[np.ndarray, float] = 0,
     scale: Union[np.ndarray, float] = 1,
-    size: int = 1,
```
Comment (toshihikoyanase) [inline optuna/samplers/_tpe/_truncnorm.py:210]: [Notes] I confirmed that `size` of the `rvs` function was not used in the feature branch.

File: optuna/samplers/_tpe/probability_distributions.py [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+    step: float
+
+
+_BatchedDistributions = Union[
+    _BatchedCategoricalDistributions,
+    _BatchedTruncNormDistributions,
+    _BatchedDiscreteTruncNormDistributions,
+]
+
+EPS = 1e-12
```
Comment (toshihikoyanase) [inline optuna/samplers/_tpe/probability_distributions.py:35]: Now that `_normal_cdf` was removed, `EPS` is not used in this file.
```suggestion
```

Comment (contramundum53) [reply optuna/samplers/_tpe/probability_distributions.py:35]: Thanks for pointing it out!
I applied the suggestion.

**Review by toshihikoyanase — APPROVED** [review]
This is actually re-submission of #4183, which had already got two approvals. I confirmed that this PR with the following two points:
- The changes does not contain unexpected diff
- `TPESampler` worked with some optuna examples

## Diff

`_truncnorm.py`: 
- `with np.errstate(divide="ignore")` was removed from #4183 and keep the code in master.
- `size` argument of `rvs` was removed since it was not used.

```diff
37d36
< import typing
40d38
< from typing import Tuple
57,58c55
<     with np.errstate(divide="ignore"):  # We ignore warnings for log(0)
<         return log_p + np.log1p(-np.exp(log_q - log_p))
---
>     return log_p + np.log1p(-np.exp(log_q - log_p))
184,185c181
<         with np.errstate(divide="ignore"):  # We ignore warnings for log(0)
<             log_Phi_x = _log_sum(_log_ndtr(a), np.log(q) + _log_gauss_mass(a, b))
---
>         log_Phi_x = _log_sum(_log_ndtr(a), np.log(q) + _log_gauss_mass(a, b))
189,190c185
<         with np.errstate(divide="ignore"):  # We ignore warnings for log(0)
<             log_Phi_x = _log_sum(_log_ndtr(-b), np.log1p(-q) + _log_gauss_mass(a, b))
---
>         log_Phi_x = _log_sum(_log_ndtr(-b), np.log1p(-q) + _log_gauss_mass(a, b))
215d209
<     size: Optional[int] = None,
218,221d211
<     if size is not None:
<         shape = typing.cast(Tuple[int, ...], (size,))
<     else:
<         shape = np.broadcast(a, b, loc, scale).shape
223c213,214
<     percentiles = random_state.uniform(low=0, high=1, size=shape)
---
>     size = np.broadcast(a, b, loc, scale).shape
>     percentiles = random_state.uniform(low=0, high=1, size=size)
```

`parzen_estimator.py`: A blank line was removed. I guess the linters took care of it.

```diff
72d71
<
```

`probability_distributions.py`:  Removed some unused lines
```diff
7d6
< from optuna.samplers._tpe import _erf
35,42d33
<
< EPS = 1e-12
<
<
< def _normal_cdf(
<     x: Union[float, np.ndarray], mu: Union[float, np.ndarray], sigma: Union[float, np.ndarray]
< ) -> np.ndarray:
<     return 0.5 * (1 + _erf.erf((x - mu) / np.maximum(np.sqrt(2) * sigma, EPS)))
```

`test_parzen_estimator.py` : None
`test_probability_distributions.py` : None

## Examples

I tried some example code in [optuna-examples](https://github.com/optuna/optuna-examples) including quadratic_simple.py and sklearn_simple.py, and the differences were not obvious. If we want to compare performance quantitatively, I guess we need to use some banchmark suites such as kurobako.

**Pull request merged**